classes/metric/base.php

```diff
@@ -154,4 +177,37 @@ public function set_last_generate_time(int $timestamp) {
      * @return metric_item
      */
     abstract public function generate_metric_item(int $starttime, int $finishtime): metric_item;
+
+    /**
+     * Metric's ability to be backfilled.
+     *
+     * @return bool
+     */
+    public function is_backfillable(): bool {
+        return false;
+    }
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param int $backwardperiod Time from which sample is to be retrieved.
+     * @param int $finishtime If data is being completed argument is passed here.
+     *
+     * @return array|null
+     */
+    public function generate_metric_items(int $backwardperiod, int $finishtime = null): ?array {
+        return null;
+    }
+
+    /**
+     * Returns two dates representing start end end log points, hence available data to retrieve.
+     *
+     * @return object|bool
+     */
+    public function get_range_log_available() {
+        global $DB;
+
+        $sql = 'SELECT max(timecreated) AS "max", min(timecreated) AS "min" FROM {logstore_standard_log}';
```

Comment:
would be slightly nicer to format this
https://docs.moodle.org/dev/SQL_coding_style#Indentation